main/uart.c

```diff
@@ -15,321 +15,451 @@
 #define BUF_MAX 1000
 
-#define STX 0x2
-#define ETX 0x3
-#define NRF_CMD1 0x1
-#define NRF_CMD2 0x2
-#define NRF_CMD1_LEN 3
-#define NRF_CMD2_LEN 1
-
-static const char *TAG = "uart";
+static const char * TAG = "uart";
 
 RingbufHandle_t uart_temp_buf;
 portMUX_TYPE adv_table_mux = portMUX_INITIALIZER_UNLOCKED;
 
 struct adv_report_table adv_reports;
 struct adv_report_table adv_reports_buf;
 
-int adv_put_to_table(adv_report_t* new_adv)
+/**
+ * @brief Print given binary into hex string.
+ *
+ * Example {0xA0, 0xBB, 0x31} -> "A0BB31"
+ *
+ * @param[out] hexstr String to print into. Must be at least 2 * binlen + 1 bytes long.
+ * @param[in]  bin Binary to print from.
+ * @param[in]  binlen Size of binary in bytes.
+ */
+static void bin2hex (char * const hexstr, const uint8_t * const bin, size_t binlen)
 {
-	portENTER_CRITICAL(&adv_table_mux);
-	bool found = false;
-	esp_err_t ret = ESP_OK;
-
-	for (int i=0; i<adv_reports.num_of_advs; i++) {
-		char* mac = adv_reports.table[i].tag_mac;
-		if (strcmp(new_adv->tag_mac, mac) == 0) {
-			found = true;
-			adv_reports.table[i] = *new_adv;
-		}
-	}
-
-	//not found from the table, insert if not full
-	if (!found) {
-		if (adv_reports.num_of_advs < MAX_ADVS_TABLE) {
-			adv_reports.table[adv_reports.num_of_advs++] = *new_adv;
-		} else {
-			ret = ESP_ERR_NO_MEM;
-		}
-	}
-	portEXIT_CRITICAL(&adv_table_mux);
-
-	return ret;
+    for (size_t ii = 0; ii < binlen; ii++)
+    {
+        sprintf (hexstr + (2 * ii), "%02X", bin[ii]);
+    }
+
+    hexstr[2 * binlen] = 0;
 }
 
-int uart_send_data(const char* logName, const char* data)
+static esp_err_t adv_put_to_table (const adv_report_t * const p_adv)
 {
-	char buf[100] = { 0 };
-	const int len = strlen(data);
+    portENTER_CRITICAL (&adv_table_mux);
+    bool found = false;
+    esp_err_t ret = ESP_OK;
+
+    // Check if we already have advertisement with this MAC
+    for (int i = 0; i < adv_reports.num_of_advs; i++)
+    {
+        char * mac = adv_reports.table[i].tag_mac;
+
+        if (strcmp (p_adv->tag_mac, mac) == 0)
+        {
+            // Yes, update data.
+            found = true;
+            adv_reports.table[i] = *p_adv;
+        }
+    }
+
+    //not found from the table, insert if not full
+    if (!found)
+    {
+        if (adv_reports.num_of_advs < MAX_ADVS_TABLE)
+        {
+            adv_reports.table[adv_reports.num_of_advs++] = *p_adv;
+        }
+        else
+        {
+            ret = ESP_ERR_NO_MEM;
+        }
+    }
+
+    portEXIT_CRITICAL (&adv_table_mux);
+    return ret;
+}
 
-	for (int i = 0; i < len; i++) {
-		sprintf(buf+(2*i), "%02x", data[i]);
-	}
+int uart_send_data (const char * logName, const char * data)
+{
+    char buf[100] = { 0 };
+    const int len = strlen (data);
 
-	const int txBytes = uart_write_bytes(UART_NUM_1, data, len);
-	ESP_LOGI(logName, "Wrote to uart %d bytes, 0x%s", txBytes, buf);
+    for (int i = 0; i < len; i++)
+    {
+        sprintf (buf + (2 * i), "%02x", data[i]);
```

Comment:
sprintf is very unsafe here, it can cause stack damaging if the length of data will be bigger than 49 bytes.
Please consider using snprintf or ESP_LOG_BUFFER_HEXDUMP

Reply:
It would be better to pass the length of data as an argument, just in case that NULL byte has ever to be sent